README.md
#  Branca Tokens for PHP

Authenticated and encrypted API tokens using modern crypto.

[![Latest Version](https://img.shields.io/packagist/v/tuupola/branca.svg?style=flat-square)](https://packagist.org/packages/tuupola/branca)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Build Status](https://img.shields.io/github/workflow/status/tuupola/branca-php/Tests/2.x?style=flat-square)](https://github.com/tuupola/branca-php/actions)
[![Coverage](https://img.shields.io/codecov/c/github/tuupola/branca-php.svg?style=flat-square)](https://codecov.io/github/tuupola/branca-php)



## What?

[Branca](https://github.com/tuupola/branca-spec) is a secure easy to use token format which makes it hard to shoot yourself in the foot. It uses IETF XChaCha20-Poly1305 AEAD symmetric encryption to create encrypted and tamperproof tokens. Payload itself is an arbitrary sequence of bytes. You can use for example a JSON object, plain text string or even binary data serialized by [MessagePack](http://msgpack.org/) or [Protocol Buffers](https://developers.google.com/protocol-buffers/).

It is possible to use [Branca as an alternative to JWT](https://appelsiini.net/2017/branca-alternative-to-jwt/). There is also an [authentication middleware](https://github.com/tuupola/branca-middleware) for frameworks which support PSR-7 doublepass or PSR-15 standards.

## Install

Install the library using [Composer](https://getcomposer.org/).


``` bash
$ composer require tuupola/branca
```

This branch requires PHP 7.2 or up. The older 1.x branch supports also PHP 5.6, 7.0 and 7.1.

``` bash
$ composer require "tuupola/branca:^1.0"
```

## Usage

Token payload can be any arbitrary data such as string containing an email
address. You also must provide a 32 byte secret key. The key is used for encrypting the payload.

```php
use Branca\Branca;

$key = random_bytes(32);
$branca = new Branca($key);

$payload = "tuupola@appelsiini.net";
$token = $branca->encode($payload);
/* hGgg0dPSseaUPZqGloWlDGb2i8hb6iamFBIQaatgYDRhEuaXyByaX0nzmyQk1WYAuSBEMWpB20Z1dENLFItwf1 */

$decoded = $branca->decode($token);
/* tuupola@appelsiini.net */
```

Sometimes you might prefer JSON.

```php
use Branca\Branca;

$key = random_bytes(32);
$branca = new Branca($key);

$payload = json_encode(["scope" => ["read", "write", "delete"]]);
$token = $branca->encode($payload);

/*
5R7p5pC1gU5kfVuBUzhl43Ndh4HLT9fxAHrhN1zNRivTuehY8zYYzrVZ8C6d6VcNLfCk3EUgBwwW6kIk0wm32O34OFIYz5LnOIezwcV2Xsfc
*/

$decoded = $branca->decode($token);
$array = json_decode($decoded, true);

/*
Array
(
    [scope] => Array
        (
            [0] => read
            [1] => write
            [2] => delete
        )

)
*/
```

You can keep the token size small by using a space efficient serialization method such as [MessagePack](http://msgpack.org/) or [Protocol Buffers](https://developers.google.com/protocol-buffers/).

```php
use Branca\Branca;
use MessagePack\MessagePack;
use MessagePack\Packer;
use MessagePack\BufferUnpacker;

$key = random_bytes(32);
$branca = new Branca($key);

$payload = (new Packer)->pack(["scope" => ["read", "write", "delete"]]);
$token = $branca->encode($payload);

/*
3iJt0CjqTRh3FGuAf0DHEmhULFIbPVInjguWIkmyCm7RMps5BMJZKa1KwZMN0z58IpPeCxdjoTdkurn9pl0YNrxAQfg3deP0
*/

$decoded = $branca->decode($token);
$unpacked = (new BufferUnpacker($decoded))->unpack();
print_r($unpacked);

/*
Array
(
    [scope] => Array
        (
            [0] => read
            [1] => write
            [2] => delete
        )

)
*/
```

## Timestamp

Branca token includes a timestamp when it was created. When decoding you can optionally pass a `ttl` parameter. Value is passed in seconds. Below example throws en exception if token is older than 60 minutes.

```php
use Branca\Branca;

$key = hex2bin("73757065727365637265746b6579796f7573686f756c646e6f74636f6d6d6974");
$branca = new Branca($key);

$token = "1jJDJOEeG2FutA8g7NAOHK4Mh5RIE8jtbXd63uYbrFDSR06dtQl9o2gZYhBa36nZHXVfiGFz";

print $branca->timestamp($token); /* 123206400 */

try {
    $decoded = $branca->decode($token, 3600);
} catch (RuntimeException $exception) {
    print $exception->getMessage(); /* Token is expired */
}
```

## Testing

You can run tests either manually or automatically on every code change. Automatic tests require [entr](http://entrproject.org/) to work.

``` bash
$ make test
```
``` bash
$ brew install entr
$ make watch
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email tuupola@appelsiini.net instead of using the issue tracker.

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.


## [2.3.0](https://github.com/tuupola/branca/compare/2.2.1...2.3.0) - 2021-06-22
### Changed
- Token is now validated before using the `timestamp()` helper ([#19](https://github.com/tuupola/branca-php/pull/19)).


## [2.2.1](https://github.com/tuupola/branca/compare/2.2.0...2.2.1) - 2020-10-29
### Fixed
- Add `SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_KEYBYTES` define which is missing from some PHP 7.2 installations ([#12](https://github.com/tuupola/branca-php/pull/12)).

## [2.2.0](https://github.com/tuupola/branca/compare/2.1.0...2.2.0) - 2020-09-10
### Added
- Allow installing with PHP 8 ([#11](https://github.com/tuupola/branca-php/pull/11)).

## [2.1.0](https://github.com/tuupola/branca/compare/2.0.0...2.1.0) - 2019-02-03
### Added
- Helper method to extract timestamp from a token ([#4](https://github.com/tuupola/branca-php/pull/4)).
- Compatibility tests from spec ([#7](https://github.com/tuupola/branca-php/pull/7)).

## [2.0.0](https://github.com/tuupola/branca/compare/1.0.0...2.0.0) - 2019-01-11
### Changed
- PHP 7.2 is now minimum requirement
- PHP bundled Sodium extension is now used
- All methods have return types
- All methods are typehinted
- All type juggling is removed

## [1.0.0](https://github.com/tuupola/branca/compare/0.3.4...1.0.0) - 2019-01-08
### Added
- Allow using tuupola/base62 ^2.0.

### Changed
- Constructor now throws `InvalidArgumentException` if invalid key is given.

## [0.3.4](https://github.com/tuupola/branca/compare/0.3.3...0.3.4) - 2018-12-09
### Added
- Allow using tuupola/base62 ^1.0.

## [0.3.3](https://github.com/tuupola/branca/compare/0.3.2...0.3.3) - 2018-09-11
### Added
- Allow using tuupola/base62 ^0.11.0.

## [0.3.2](https://github.com/tuupola/branca/compare/0.3.1...0.3.2) - 2018-04-05
### Added
- Allow using tuupola/base62 ^0.10.0.

## [0.3.1](https://github.com/tuupola/branca/compare/0.3.0...0.3.1) - 2017-12-12
### Fixed
- Token validation intermittently failed if nonce contained null bytes.

### Added
- Allow using both tuupola/base62 ^0.8.0 and ^0.9.0.

## [0.3.0](https://github.com/tuupola/branca/compare/0.2.0...0.3.0) - 2017-07-23
### Changed
- Use more secure IETF XChaCha20-Poly1305 AEAD encryption.

## [0.2.0](https://github.com/tuupola/branca/compare/0.1.0...0.2.0) - 2017-07-21
### Changed
-  Use more developer friendly unsigned 32 bit second timestamps.

## 0.1.0 - 2017-07-20

Initial realease using IETF ChaCha20-Poly1305 AEAD and 64 bit microsecond timestamps.


tests/BrancaTest.php
<?php

declare(strict_types = 1);

/*

Copyright (c) 2017-2021 Mika Tuupola

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

/**
 * @see       https://branca.io/
 * @see       https://github.com/tuupola/branca-php
 * @see       https://github.com/tuupola/branca-spec
 * @license   https://www.opensource.org/licenses/mit-license.php
 */

namespace Branca;

use InvalidArgumentException;
use RuntimeException;
use SodiumException;
use PHPUnit\Framework\TestCase;
use Tuupola\Base62;
use Nyholm\NSA;

class BrancaTest extends TestCase
{
    /**
     * Decoding tests are be mandatory for all implementations. They make
     * sure the implementation can consume tokens with different kind of edge
     * cases.
     */

    /**
      * Test vector 8
      */
    public function testShouldDecodeHelloWorldWithZeroTimestamp()
    {
        $token = "870S4BYxgHw0KnP3W9fgVUHEhT5g86vJ17etaC5Kh5uIraWHCI1psNQGv298ZmjPwoYbjDQ9chy2z";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $payload = $branca->decode($token);

        $this->assertEquals("Hello world!", $payload);
        $this->assertEquals(0, $branca->timestamp($token));
    }

    /**
      * Test vector 9
      */
    public function testShouldDecodeHelloWorldWithMaxTimestamp()
    {
        $token = "89i7YCwu5tWAJNHUDdmIqhzOi5hVHOd4afjZcGMcVmM4enl4yeLiDyYv41eMkNmTX6IwYEFErCSqr";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);

        $this->assertEquals("Hello world!", $decoded);
        $this->assertEquals(4294967295, $branca->timestamp($token));
    }
    /**
      * Test vector 10
      */
    public function testShouldDecodeHelloWorldWithNov27Timestamp()
    {
        $token = "875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);

        $this->assertEquals("Hello world!", $decoded);
        $this->assertEquals(123206400, $branca->timestamp($token));
    }

    /**
      * Test vector 11
      */
    public function testShouldDecodeEightNullBytesWithZeroTimestamp()
    {
        $token = "1jIBheHbDdkCDFQmtgw4RUZeQoOJgGwTFJSpwOAk3XYpJJr52DEpILLmmwYl4tjdSbbNqcF1";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);

        $this->assertEquals("0000000000000000", bin2hex($decoded));
        $this->assertEquals(0, $branca->timestamp($token));
    }

    /**
      * Test vector 12
      */
    public function testShouldDecodeEightNullBytesWithMaxTimestamp()
    {
        $token = "1jrx6DUu5q06oxykef2e2ZMyTcDRTQot9ZnwgifUtzAphGtjsxfbxXNhQyBEOGtpbkBgvIQx";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);

        $this->assertEquals("0000000000000000", bin2hex($decoded));
        $this->assertEquals(4294967295, $branca->timestamp($token));
    }

    /**
      * Test vector 13
      */
    public function testShouldDecodeEightNullBytesWithNov27Timestamp()
    {
        $token = "1jJDJOEjuwVb9Csz1Ypw1KBWSkr0YDpeBeJN6NzJWx1VgPLmcBhu2SbkpQ9JjZ3nfUf7Aytp";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);

        $this->assertEquals("0000000000000000", bin2hex($decoded));
        $this->assertEquals(123206400, $branca->timestamp($token));
    }

    /**
      * Test vector 14
      */
    public function testShouldDecodeEmptyPayloadWithZeroTimestamp()
    {
        $token = "4sfD0vPFhIif8cy4nB3BQkHeJqkOkDvinI4zIhMjYX4YXZU5WIq9ycCVjGzB5";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $payload = $branca->decode($token);

        $this->assertEquals("", $payload);
        $this->assertEquals(0, $branca->timestamp($token));
    }

    /**
      * Test vector 15
      */
    public function testShouldDecodeNonUtf8CharactersWithNov27Timestamp()
    {
        $token = "K9u6d0zjXp8RXNUGDyXAsB9AtPo60CD3xxQ2ulL8aQoTzXbvockRff0y1eXoHm";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $payload = $branca->decode($token);

        $this->assertEquals("80", bin2hex($payload));
        $this->assertEquals(123206400, $branca->timestamp($token));
    }

    /**
     * Test vector 16
     * Token is otherwise valid, but it has version 0xBB.
     */
    public function testShouldThrowWithWrongVersion()
    {
        $this->expectException(RuntimeException::class);

        $token = "89mvl3RkwXjpEj5WMxK7GUDEHEeeeZtwjMIOogTthvr44qBfYtQSIZH5MHOTC0GzoutDIeoPVZk3w";

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
      * Test vector 17
      */
    public function testShouldThrowWithInvalidBase62Characters()
    {
        $this->expectException(InvalidArgumentException::class);

        $token = "875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT_";

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
     * Test vector 18
     *
     * Token was created with version: 0xba
     * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT
     *
     * Before base62 encoding the version was modified to: 0xbb
     * 89mvl3S0BE0UCMIY94xxIux4eg1w5oXrhvCEXrDAjusSbO0Yk7AU6FjjTnbTWTqogLfNPJLzecHVb
     */
    public function testShouldThrowWithModifiedVersion()
    {
        $this->expectException(RuntimeException::class);

        $token = "89mvl3S0BE0UCMIY94xxIux4eg1w5oXrhvCEXrDAjusSbO0Yk7AU6FjjTnbTWTqogLfNPJLzecHVb";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
      * Test vector 19
      *
      * Token was created with nonce:
      * hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef")
      * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT
      *
      * Before base62 encoding the nonce was modified to:
      * hex2bin("00efbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef")
      * 875GH233SUysT7fQ711EWd9BXpwOjB72ng3ZLnjWFrmOqVy49Bv93b78JU5331LbcY0EEzhLfpmSx
      */
    public function testShouldThrowWithModifiedNonce()
    {
        $this->expectException(RuntimeException::class);

        $token = "875GH233SUysT7fQ711EWd9BXpwOjB72ng3ZLnjWFrmOqVy49Bv93b78JU5331LbcY0EEzhLfpmSx";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
     * Test vector 20
     *
     * Token was created with time: hex2bin("0757fb00") ie. 123206400
     * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT
     *
     * Before base62 encoding the time was modified to: hex2bin("0057fb00") ie. 5765888
     * 870g1RCk4lW1YInhaU3TP8u2hGtfol16ettLcTOSoA0JIpjCaQRW7tQeP6dQmTvFIB2s6wL5deMXr
     */
    public function testShouldThrowWithModifiedTimestamp()
    {
        $this->expectException(RuntimeException::class);

        $token = "870g1RCk4lW1YInhaU3TP8u2hGtfol16ettLcTOSoA0JIpjCaQRW7tQeP6dQmTvFIB2s6wL5deMXr";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
     * Test vector 21
     *
     * Original ciphertext in token was:
     * hex2bin("d8fdbaf35dc37a98b523e6fe")
     * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT
     *
     * Before base62 encoding the ciphertext was modified to:
     * hex2bin("d8fdbaf35dc37a98b523e600")
     * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5Qw6Jpo96myliI3hHD7VbKZBYh
     */
    public function testShouldThrowWithModifiedCiphertext()
    {
        $this->expectException(RuntimeException::class);

        $token = "875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5Qw6Jpo96myliI3hHD7VbKZBYh";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
     * Test vector 22
     *
     * Original Poly1305 tag was:
     * hex2bin("f3faf98dd385c68046fb7ed63c94995b")
     * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT
     *
     * Before base62 encoding the Poly1305 tag was modified to:
     * hex2bin("f3faf98dd385c68046fb7ed63c949900")
     * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trk0
     */
    public function testShouldThrowWithModifiedPoly1305Tag()
    {
        $this->expectException(RuntimeException::class);

        $token = "875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trk0";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
     * Test vector 23
     */
    public function testShouldThrowWithWrongKey()
    {
        $this->expectException(RuntimeException::class);

        $token = "870S4BYxgHw0KnP3W9fgVUHEhT5g86vJ17etaC5Kh5uIraWHCI1psNQGv298ZmjPwoYbjDQ9chy2z";

        $branca = new Branca("wrongsecretkeyyoushouldnotcommit");
        $decoded = $branca->decode($token);
    }

    /**
     * Test vector 24
     *
     * Key length should be 32 bytes.
     */
    public function testShouldThrowWithInvalidKey()
    {
        $this->expectException(InvalidArgumentException::class);

        $token = "870S4BYxgHw0KnP3W9fgVUHEhT5g86vJ17etaC5Kh5uIraWHCI1psNQGv298ZmjPwoYbjDQ9chy2z";

        $branca = new Branca("tooshortkey");
        $decoded = $branca->decode($token);
    }

    /**
     * Encoding tests should be implemented if the implementation allows user provided
     * nonce for unit tests. User provided nonce should *not* be part of public API
     * since it is a footgun.
     */

    /**
     * Test vector 0
     */
    public function testShouldEncodeHelloWorldWithZeroTimestamp()
    {
        $token = "870S4BYxgHw0KnP3W9fgVUHEhT5g86vJ17etaC5Kh5uIraWHCI1psNQGv298ZmjPwoYbjDQ9chy2z";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = "Hello world!";
        $timestamp = 0;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
     * Test vector 1
     */
    public function testShouldEncodeHelloWorldWithMaxTimestamp()
    {
        $token = "89i7YCwu5tWAJNHUDdmIqhzOi5hVHOd4afjZcGMcVmM4enl4yeLiDyYv41eMkNmTX6IwYEFErCSqr";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = "Hello world!";
        $timestamp = 4294967295;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
     * Test vector 2
     */
    public function testShouldEncodeHelloWorldWithNov27Timestamp()
    {
        $token = "875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = "Hello world!";
        $timestamp = 123206400;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
     * Test vector 3
     */
    public function testShouldEncodeEightNullBytesWithZeroTimestamp()
    {
        $token = "1jIBheHbDdkCDFQmtgw4RUZeQoOJgGwTFJSpwOAk3XYpJJr52DEpILLmmwYl4tjdSbbNqcF1";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = hex2bin("0000000000000000");
        $timestamp = 0;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
     * Test vector 4
     */
    public function testShouldEncodeEightNullBytesWithMaxTimestamp()
    {
        $token = "1jrx6DUu5q06oxykef2e2ZMyTcDRTQot9ZnwgifUtzAphGtjsxfbxXNhQyBEOGtpbkBgvIQx";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = hex2bin("0000000000000000");
        $timestamp = 4294967295;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
     * Test vector 5
     */
    public function testShouldEncodeEightNullBytesWithNov27Timestamp()
    {
        $token = "1jJDJOEjuwVb9Csz1Ypw1KBWSkr0YDpeBeJN6NzJWx1VgPLmcBhu2SbkpQ9JjZ3nfUf7Aytp";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = hex2bin("0000000000000000");
        $timestamp = 123206400;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
     * Test vector 6
     */
    public function testShouldEncodeEmptyPayloadWithZeroTimestamp()
    {
        $token = "4sfD0vPFhIif8cy4nB3BQkHeJqkOkDvinI4zIhMjYX4YXZU5WIq9ycCVjGzB5";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = "";
        $timestamp = 0;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
      * Test vector 7
      */
    public function testShouldEncodeNonUtf8Characters()
    {
        $token = "K9u6d0zjXp8RXNUGDyXAsB9AtPo60CD3xxQ2ulL8aQoTzXbvockRff0y1eXoHm";
        $nonce = hex2bin("beefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef");
        $payload = hex2bin("80");
        $timestamp = 123206400;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload, $timestamp);

        $this->assertEquals($token, $encoded);
    }

    /**
     * These are PHP implementation specific tests. For example TTL check is not a
     * feature of the token, it is feature of the implementation.
     */
    public function testShouldThrowWhenTtlExpired()
    {
        $this->expectException(RuntimeException::class);

        $timestamp = 123206400;

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $token = $branca->encode("Hello world!", $timestamp);
        $decoded = $branca->decode($token, 3600);
    }

    public function testShouldThrowWithInvalidNonce()
    {
        $this->expectException(SodiumException::class);

        $nonce = hex2bin("dead");
        $payload = "Hello world!";

        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        NSA::setProperty($branca, "nonce", $nonce);

        $encoded = $branca->encode($payload);
    }

    /**
     * Timestamp helper should throw is token has been tampered.
     *
     * Token was created with time: hex2bin("0757fb00") ie. 123206400
     * 875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT
     *
     * Before base62 encoding the time was modified to: hex2bin("0057fb00") ie. 5765888
     * 870g1RCk4lW1YInhaU3TP8u2hGtfol16ettLcTOSoA0JIpjCaQRW7tQeP6dQmTvFIB2s6wL5deMXr
     */
    public function testShouldThrowWithModifiedTimestampWhenReadingTimestamp()
    {
        $this->expectException(RuntimeException::class);

        $token = "870g1RCk4lW1YInhaU3TP8u2hGtfol16ettLcTOSoA0JIpjCaQRW7tQeP6dQmTvFIB2s6wL5deMXr";
        $branca = new Branca("supersecretkeyyoushouldnotcommit");
        $timestamp = $branca->timestamp($token);
    }

    // public function testShouldThrowWhenTimestampOverflows()
    // {
    //     $this->expectException(RuntimeException::class);

    //     /* Token with 123206400 timestamp. */
    //     $token = "875GH23U0Dr6nHFA63DhOyd9LkYudBkX8RsCTOMz5xoYAMw9sMd5QwcEqLDRnTDHPenOX7nP2trlT";

    //     $branca = new Branca("supersecretkeyyoushouldnotcommit");

    //     /* Add maximum value to make timestamp overflow. */
    //     $ttl = 4294967295;
    //     $decoded = $branca->decode($token, $ttl);
    //}
}


src/Branca.php
<?php

declare(strict_types = 1);

/*

Copyright (c) 2017-2021 Mika Tuupola

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

/**
 * @see       https://branca.io/
 * @see       https://github.com/tuupola/branca-php
 * @see       https://github.com/tuupola/branca-spec
 * @license   https://www.opensource.org/licenses/mit-license.php
 */

namespace Branca;

use Tuupola\Base62;

class Branca
{
     /**
      *  Magic byte, BrancA.
      */
    const VERSION = 0xBA;

    /**
     * @var string
     */
    private $key;

    /**
     * @var int
     */
    private $timestamp = null;

    /**
     * Nonce used for unit testing only.
     * @var string
     */
    private $nonce = null;


    public function __construct(string $key)
    {
        /* Apparently some PHP 7.2 versions do not have this defined. */
        if (!defined("SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_KEYBYTES")) {
            define("SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_KEYBYTES", 32);
        }

        if (SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_KEYBYTES !== strlen($key)) {
            throw new InvalidKeyException(
                sprintf("Key must be exactly %d bytes", SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES)
            );
        }
        $this->key = $key;
    }

    /**
     * Create a token from payload and optional timestamp.
     */
    public function encode(string $payload, int $timestamp = null): string
    {
        if (null === $timestamp) {
            $timestamp = time();
        }

        $version = pack("C", self::VERSION);
        $time = pack("N", $timestamp);

        $nonce = $this->nonce;
        if (empty($nonce)) {
            $nonce = random_bytes(SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES);
        }

        $header = $version . $time . $nonce;

        $ciphertext = sodium_crypto_aead_xchacha20poly1305_ietf_encrypt(
            $payload,
            $header,
            $nonce,
            $this->key
        );

        $token =  $header . $ciphertext;
        return (new Base62)->encode($token);
    }

    /**
     * Extract the payload from a token.
     */
    public function decode(string $token, int $ttl = null): string
    {
        $token = (new Base62)->decode($token);
        $header = substr(
            $token,
            0,
            SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES + 5
        );
        $nonce = substr(
            $header,
            5,
            SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES
        );
        $ciphertext = substr(
            $token,
            SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES + 5
        );

        $parts = unpack("Cversion/Ntime", $header);

        /* Unpack failed, should not ever happen. */
        if (false === $parts) {
            throw new InvalidTokenException("Cannot extract token header");
        }

        /* Implementation should accept only current version. */
        if ($parts["version"] !== self::VERSION) {
            throw new InvalidTokenVersionException("Invalid token version");
        }

        try {
            $payload = sodium_crypto_aead_xchacha20poly1305_ietf_decrypt(
                $ciphertext,
                $header,
                $nonce,
                $this->key
            );
        } catch (\Throwable $error) {
            throw new InvalidTokenException("Invalid token");
        /** @phpstan-ignore-next-line */
        } catch (\Exception $error) {
            throw new InvalidTokenException("Invalid token");
        }

        /* In some cases sodium returns false. */
        if (false === $payload) {
            throw new InvalidTokenException("Invalid token");
        }

        /* Store timestamp value for the helper. */
        $this->timestamp = $parts["time"];

        /* Check for expired token if TTL is set. */
        if (is_integer($ttl)) {
            $future = $parts["time"] + $ttl;
            if ($future < time()) {
                throw new ExpiredTokenException("Token is expired");
            }
        }

        return $payload;
    }

    public function timestamp(string $token): int
    {
        if (null === $this->timestamp) {
            $this->decode($token);
        }

        return $this->timestamp;
    }
}
